### PR TITLE
Terraform variables closes 425

### PR DESCRIPTION
**Pull request by tvaidotas**

### CODE REVIEW

File: topics/terraform/modules/variables/README.md [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+	region = "eu-west-2"
+}
+``` 
+
+#### Adding resource
+
+Paste the following below the variable in the `main.tf` file:
+```hcl
+resource "aws_instance" "example" {
+	ami = ami
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
	ami = var.ami
```

File: topics/terraform/modules/variables/README.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+	instance_type = type
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
	instance_type = var.type
```

File: topics/terraform/modules/variables/README.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+`main.tf` configuration file should look like this:
+
+```hcl
+provider "aws" {
+  region     = "eu-west-2"
+}
+
+resource "aws_instance" "example" {
+  ami           = ami
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
  ami           = var.ami
```

File: topics/terraform/modules/variables/README.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+  instance_type = type
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
  instance_type = var.type
```

File: topics/terraform/modules/variables/README.md [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+Lastly let's create the resource by executing:
+
+`terraform apply -var-file="variableValues.tfvars"`
+
+Once terraform will give you a prompt about the successful operation in the *AWS console* under *Compute* and then *EC2* check that the resource has been created. 
+
+Make sure that you are within the correct region, otherwise you won't be able to see the resource.
+
+#### Overview
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
#### Summary
```

File: topics/terraform/modules/variables/README.md [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+- then environment variables
+- followed by automatic variable files
+- and -var and `-var-files` share the highest precedence level
+
+When variables are defined within the same precedence level, the last value is used. 
+For example, if you define a variable in a var-file and a -var option, the one that you specify last on the command-line will take precedence.
+
+## Tasks
+
+### AWS task 1
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
### AWS Variables with TFVARS File
```

File: topics/terraform/modules/variables/README.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+To delete the created resource run the following command in the terminal, make sure that the terminal is in the directory where `main.tf` is located:
+`terraform destroy` 
+
+Check in the *AWS console* under *Compute* and then *EC2* check that the resource has been deleted.
+
+Make sure that you are within the correct region, otherwise you won't be able to see the resource.
+
+</details>
+
+### AWS task 2
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
### AWS Variables with CLI Option
```

File: topics/terraform/modules/variables/README.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+<details>
+
+<summary>In this task you will attempt to provide variable values in a different way.</summary>
+
+#### Prerequisites
+
+In order to do this task you need to have *Task 1* completed and have the `main.tf`, `variables.tf` files.
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
In order to do this task you need to have the *AWS Variables with TFVARS File* (previous) task completed and have the `main.tf`, `variables.tf` files.
```

File: topics/terraform/modules/variables/README.md [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+#### Providing values through the terminal
+
+Create a new directory like `example_5`.
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
Create a new directory like `terraform-variables-aws-using-cli`.
```

File: topics/terraform/modules/variables/README.md [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+
+Lastly let's create the resource by executing, you will be prompted to provide variable values as well once you execute the following command:
+
+`terraform apply`
+
+Once terraform will give you a prompt about the successful operation in the *AWS console* under *Compute* and then *EC2* check that the resource has been created. 
+
+Make sure that you are within the correct region, otherwise you won't be able to see the resource.
+
+#### Overview
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
#### Summary
```

File: topics/terraform/modules/variables/README.md [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+In this task you used three configuration files to deploy a resource in AWS.
+
+Additionally the variable values were defined in a different file from where the variables were declared.
+
+#### Clean up
+
+To delete the created resource run the following command in the terminal, make sure that the terminal is in the directory where `main.tf` is located:
+`terraform destroy`
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
`terraform destroy -var-file="variableValues.tfvars"` 
```

File: topics/terraform/modules/variables/README.md [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+First let's execute the following command to get the plugins for AWS:
+
+`terraform init`
+
+Next let's execute to see what changes will be made, as there's no `variablesValues.tfvars` file those values do need to come from somewhere. 
+Once you're going to execute the following command, you will be prompted to provide values for the variables. 
+A name of the variable will be displayed, similarly if there is a `description` it will be printed to you as well in order to give some context.
+
+`terraform plan`
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
`terraform plan -var='ami=ami-f976839e' -var='type=t2.micro'`
```

File: topics/terraform/modules/variables/README.md [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+
+Lastly let's create the resource by executing, you will be prompted to provide variable values as well once you execute the following command:
+
+`terraform apply`
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
`terraform apply -var='ami=ami-f976839e' -var='type=t2.micro'`
```

File: topics/terraform/modules/variables/README.md [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+Make sure that you are within the correct region, otherwise you won't be able to see the resource.
+
+#### Overview
+
+In this task you provided variable values directly in the terminal when you were prompted for them.
+
+#### Clean up
+
+To delete the created resource run the following command in the terminal, make sure that the terminal is in the directory where `main.tf` is located:
+`terraform destroy`
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]:
```suggestion
`terraform destroy -var='ami=ami-f976839e' -var='type=t2.micro'` 
```

File: topics/terraform/modules/variables/README.md [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+Check in the *AWS console* under *Compute* and then *EC2* check that the resource has been deleted.
+
+Make sure that you are within the correct region, otherwise you won't be able to see the resource.
+
+</details>
+
+### Additional tasks
+
+- pass the variable values by using the `-var` flag
+- pass the variable value through environment variables
```
Comment (bob-crutchley) [inline topics/terraform/modules/variables/README.md]: Consider creating exercises for these.
This is a pretty new concept, just grab me for a chat about it.

File: topics/terraform/modules/tf_configuration_file_discovery/README.md [hunk 16]
```diff
@@ -0,0 +1,206 @@
+# Configuration file discovery
```
[the remark this answers is not in the document]
Comment (tvaidotas) [reply topics/terraform/modules/tf_configuration_file_discovery/README.md]: not really

File: topics/terraform/modules/variables/README.md [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+Variable files can be included using the var-file option or they can be automatically included if they are in your current working directory and are either named *terraform.tfvars* or have a file extension of *.auto.tfvars*. 
+You can use multiple variable files. 
+Variables defined using the `-var-file` option override variables defined using automatic variable files.
+- using environment variables. 
+When setting Terraform variables using environment variables the name of the environment variable must begin with **TF_VAR_** and be followed by the name of the Terraform variable. 
+
+## Variable Precedence
+
+When a variable is defined multiple times, the value of the variable is usually set to the highest precedent definition. 
+One thing to note is that map variables are set to the highest precedent definition.
```
[the remark this answers is not in the document]
Comment (tvaidotas) [reply topics/terraform/modules/variables/README.md]: No it's refering that by default maps are always using highest precedence

File: topics/terraform/modules/variables/README.md [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+When variables are defined within the same precedence level, the last value is used. 
+For example, if you define a variable in a var-file and a -var option, the one that you specify last on the command-line will take precedence.
+
+## Tasks
+
+### AWS task 1
+
+<details>
+
+<summary>In this task you will use a *variablesValues* file to give values to your *variables* while creating a resource on AWS.</summary>
```
[the remark this answers is not in the document]
Comment (tvaidotas) [reply topics/terraform/modules/variables/README.md]: keep forgetting

File: topics/terraform/modules/variables/README.md (hunk 12 above)
Comment (tvaidotas) [reply topics/terraform/modules/variables/README.md]: this is part of the additional task passing them in this way

File: topics/terraform/modules/variables/README.md (hunk 13 above)
Comment (tvaidotas) [reply topics/terraform/modules/variables/README.md]: this is part of the additional task passing them in this way

File: topics/terraform/modules/variables/README.md (hunk 14 above)
Comment (tvaidotas) [reply topics/terraform/modules/variables/README.md]: this is part of the additional task passing them in this way

**Pull request closed without merging**